rpcs3/Emu/RSX/rsx_cache.h

```diff
@@ -505,39 +513,97 @@ namespace rsx
 				std::vector<u8> bytes;
 				fs::file f(filename);
-
-				processed++;
-				dlg->update_msg(processed, entry_count);
-
 				if (f.size() != sizeof(pipeline_data))
 				{
 					LOG_ERROR(RSX, "Cached pipeline object %s is not binary compatible with the current shader cache", tmp.name.c_str());
 					invalid_entries.push_back(filename);
 					continue;
 				}
-
 				f.read<u8>(bytes, f.size());
+
 				auto unpacked = unpack(*(pipeline_data*)bytes.data());
-				m_storage.add_pipeline_entry(std::get<1>(unpacked), std::get<2>(unpacked), std::get<0>(unpacked), std::forward<Args>(args)...);
+				m_storage.preload_programs(std::get<1>(unpacked), std::get<2>(unpacked));
+				unpackeds.push_back(unpacked);
 
-				tally += delta;
-				if (tally > 1.f)
+				// Only update the screen at about 10fps since updating it everytime slows down the process
+				std::chrono::milliseconds now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+				processed_since_last_update++;
+				if ((now - last_update > 100ms) || (i == entry_count - 1))
+				{
+					dlg->update_msg(0, i + 1, entry_count);
+					dlg->inc_value(0, processed_since_last_update);
+					last_update = now;
+					processed_since_last_update = 0;
+				}
+			}
+
+			atomic_t<u32> processed(0);
+			std::function<void(u32)> shader_comp_worker = [&](u32 index) {
+				u32 pos;
+				while (((pos = ++processed - 1) < entry_count) && !Emu.IsStopped())
+				{
+					auto unpacked = unpackeds[pos];
+					m_storage.add_pipeline_entry(std::get<1>(unpacked), std::get<2>(unpacked), std::get<0>(unpacked), std::forward<Args>(args)...);
+				}
+			};
+
+			if (g_cfg.video.renderer == video_renderer::vulkan)
+			{
+				// Start workers
+				for (u32 i = 0; i < nb_threads; i++)
+				{
+					worker_threads[i] = std::thread(shader_comp_worker, i);
+				}
+
+				// Wait for the workers to finish their task while updating UI
+				u32 current_progress = 0;
+				u32 last_update_progress = 0;
+				do
+				{
+					std::this_thread::sleep_for(100ms); // Around 10fps should be good enough
+
+					current_progress = processed.load();
+
+					dlg->update_msg(1, current_progress, entry_count);
+					dlg->inc_value(1, current_progress - last_update_progress);
+					last_update_progress = current_progress;
+				} while ((current_progress < entry_count) && !Emu.IsStopped());
+
+				// Need to join the threads to be absolutely sure shader compilation is done.
+				for (std::thread& worker_thread : worker_threads)
+					worker_thread.join();
+			}
+			else
+			{
+				u32 pos;
+				while (((pos = ++processed - 1) < entry_count) && !Emu.IsStopped())
 				{
-					u32 value = (u32)tally;
-					dlg->inc_value(value);
+					auto unpacked = unpackeds[pos];
+					m_storage.add_pipeline_entry(std::get<1>(unpacked), std::get<2>(unpacked), std::get<0>(unpacked), std::forward<Args>(args)...);
 
-					tally -= (f32)value;
+					dlg->update_msg(1, processed.load(), entry_count);
```

Comment:
You should also do the timed updates for OGL. I was going to add this at some point but then other stuff got in the way. A huge chunk of the performance bump comes from the timed update that does not invoke the vblank wait in the driver especially on nvidia where frametimes are not uncapped by default.